### PR TITLE
Override HOME if its set to the empty string

### PR DESCRIPTION
**Pull request by rata**
I'm testing runc 1.3 in docker CI and I found these issues.

I've also added some e2e tests and verified that runc 1.2 behaves the same way.

Below the commit msg of the fix, but see each commit msg for more details

---

Before commit 06f1e0765 ("libct: speedup process.Env handling") we were
overriding HOME if it was set to "" too[1]. But now we only override it
if it wasn't set at all.

This patch restores the old behavior of overriding it if it was set to
an empty value.
    
Docker relies on this behaviour since ages[2].
    
[1]: https://github.com/opencontainers/runc/blob/1c508045727231e7342e258ab30add1478c1f981/libcontainer/init_linux.go#L544-L549
[2]: https://github.com/moby/moby/blob/843e51459f14ebc964d349eba1013dc8a3e9d52e/integration-cli/docker_cli_run_test.go#L822-L843

### CODE REVIEW

File: libcontainer/env.go [hunk 1]
```diff
@@ -49,17 +52,21 @@ func prepareEnv(env []string, uid int) ([]string, error) {
...
-			if err := os.Setenv("PATH", kv[i+1:]); err != nil {
+			if err := os.Setenv("PATH", val); err != nil {
 				return nil, err
 			}
 		}
+		// We don't override HOME if it is set to a non-empty value.
+		if key == "HOME" && val != "" {
+			overrideHome = false
+		}
 		out = append(out, kv)
```
Comment (kolyshkin) [inline libcontainer/env.go:63]: There's a bug in here. You are appending empty `HOME=` to `out`, and later you append another `HOME=`.

Should be something like
```go
if key == "HOME" {
  if val == "" {
     // For backward compatibility, ignore empty HOME.
     saw[key] = false
     continue
  }
  overrideHome = false
}
```

File: tests/integration/env.bats [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	setup_busybox
+}
+
+function teardown() {
+	teardown_bundle
+}
+
+@test "non-empty HOME env is used" {
+	update_config ' .process.env += ["HOME=/override"]'
+	update_config ' .process.args += ["-c", "echo $HOME"]'
```
Comment (kolyshkin) [inline tests/integration/env.bats:18]: nit: can merge into single `update_config`

File: tests/integration/env.bats [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+@test "env var with new-line is honored" {
+	# shelcheck: disable=SC2016
+	update_config ' .process.env += ["NEW_LINE_ENV=\n"]'
+	update_config ' .process.args += ["-c", "test -n \"$NEW_LINE_ENV\"; echo result: \"$?\""]'
+
+	runc run test_busybox
+	[ "$status" -eq 0 ]
+	# test returns 0 if the variable is set
+	[[ "${lines[0]}" == *"result: 0"* ]]
```
Comment (kolyshkin) [inline tests/integration/env.bats:51]: this is kinda redundant, you can just have

	update_config ' .process.args += ["-c", "test -n \"$NEW_LINE_ENV\""]'
	...
	[ "$status" -eq 0 ]

To check what `test` returns.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
WRT tests that you add -- maybe it's easier to write those in Go and add to  `libcontainer/env_test.go`, so all tests are in one place.

File: libcontainer/env.go (hunk 1 above)
Comment (rata) [reply libcontainer/env.go:63]: Yeah, it's not a problem to have a var several times, the last value is honored by the shell.

But sure, we can avoid it. I added only a continue in that case instead of that, we don't need to unset saw[key], we don't need to check home again, the last occurrence is the only relevant one.

File: tests/integration/env.bats (hunk 2 above)
Comment (rata) [reply tests/integration/env.bats:18]: Yeap, but I find it kind of painful when you need to modify it. IMHO this is more readable.

Do you think it is ok to keep it like this?

File: tests/integration/env.bats (hunk 3 above)
Comment (rata) [reply tests/integration/env.bats:51]: It feels like being a little too smart. Because then you need to test for 1 in this test-case, but you need to do something completely different (do what I'm doing here): https://github.com/opencontainers/runc/pull/4704/files#diff-97f18cf2c1c95327cbf79009b197aa65dff16e7a0299b2f4894d3a69fe49abceR70. 

But if you do the same and test status to be 1, you might be covering issues that runc failed to run, returned one. The test will pass in that case.

I think this reflects exactly what I want to test: the env var is set and it's quite simple anyways.

What do you think?

**Comment by rata** [comment]
@kolyshkin I thought about adding tests there. The unit tests are nice, but in a big refactor they might not exist anymore. In fact, `prepareEnv()` didn't exist before the last refactor, it might not exist after the next one.

I really think we need e2e tests for this. I want us to be able to refactor and change heavily the way we set env vars, and have tests that catch any buggy behavior. This is what the e2e test I added try to do.

**Comment by rata** [comment]
This indeed fixes docker CI and it's ready for review now :)

File: libcontainer/env.go [hunk 4]
```diff
@@ -49,17 +52,26 @@ func prepareEnv(env []string, uid int) ([]string, error) {
...
-			if err := os.Setenv("PATH", kv[i+1:]); err != nil {
+			if err := os.Setenv("PATH", val); err != nil {
 				return nil, err
 			}
 		}
+		if key == "HOME" {
+			if val != "" {
+				homeIsSet = true
+			}
+			if val == "" {
```
Comment (kolyshkin) [inline libcontainer/env.go:63]: nit: use `else` here?

File: libcontainer/env.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/env.go]: > Yeah, it's not a problem to have a var several times, the last value is honored by the shell.

If that's not a problem, when why are we performing a deduplication at all? Why Go stdlib does it?

I researched this a little when working on prepareEnv, and from what I remember we should not really trust that shell behavior because it's not quite standard (meaning, it's not really documented). This is the reason why we do the deduplication. Something somewhere could break.

> we don't need to unset saw[key], we don't need to check home again, the last occurrence is the only relevant one.

Here's a test to demonstrate why `saw[key] = false` is needed. Apply this patch to your code:

```diff
diff --git a/libcontainer/env_test.go b/libcontainer/env_test.go
index c9174716..c26b29bd 100644
--- a/libcontainer/env_test.go
+++ b/libcontainer/env_test.go
@@ -29,6 +29,10 @@ func TestPrepareEnv(t *testing.T) {
                        env:     []string{"HOME=/whoo", "FOO=bar"},
                        wantEnv: []string{"HOME=/whoo", "FOO=bar"},
                },
+               {
+                       env:     []string{"HOME=/doo", "HOME="},
+                       wantEnv: []string{"HOME=/doo"},
+               },
                {
                        env:     []string{"A=a", "A=b", "A=c"},
                        wantEnv: []string{"A=c", home},
```

and see what happens:

```console
[kir@kir-tp1 libcontainer]$ go test -v -run PrepareEnv
=== RUN   TestPrepareEnv
    env_test.go:53: want [HOME=/doo], got [HOME=/home/kir]
--- FAIL: TestPrepareEnv (0.00s)
FAIL
exit status 1
FAIL	github.com/opencontainers/runc/libcontainer	0.002s
```

File: tests/integration/env.bats (hunk 2 above)
Comment (kolyshkin) [reply tests/integration/env.bats:18]: Fine with me.

File: tests/integration/env.bats (hunk 3 above)
Comment (kolyshkin) [reply tests/integration/env.bats:63]: > But if you do the same and test status to be 1, you might be covering issues that runc failed to run, returned one. The test will pass in that case.

This is a strange argument. Say, if you want to check that variable is empty, you use `test -z`. If you want to check that variable is unset, you use `test -v`, and so on -- so that in normal case the return code is 0.

If there is a need to return non-zero exit code and check for that, I would use a unique exit code, which is never normally returned by runc. For example, like here:
https://github.com/opencontainers/runc/blob/c3a41d77db71d9ba8bd76cdcd74d8aa71d16d2a8/tests/integration/exec.bats#L31-L32

Overall, I suggested it because what your code does it converts the integer exit code into a string and then you check the string, and all this is redundant. You can argue this is more explicit, but I'm not sure it benefits anything.

But it's a very minor thing, and you can leave it as is.

**Comment by kolyshkin** [comment]
> @kolyshkin I thought about adding tests there. The unit tests are nice, but in a big refactor they might not exist anymore. In fact, `prepareEnv()` didn't exist before the last refactor, it might not exist after the next one.

In that case, the tests should migrate to a different place.

But if you modify a function which already have a unit test, it's best to amend the unit test as well.

Feel free to add bats tests, too; I understand their benefit. Just don't abandon the unit test.

File: tests/integration/env.bats [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+	[[ "${lines[0]}" == *"ONE=three"* ]]
+}
+
+@test "empty env var makes it unset" {
+	update_config ' .process.env += ["ONE=", "ONE=two", "ONE="]'
+	update_config ' .process.args += ["-c", "test -n \"$ONE\"; echo result: \"$?\"; echo ONE=X\"$ONE\"X"]'
+
+	runc run test_busybox
+	[ "$status" -eq 0 ]
+	# test returns 1 if the variable is not set.
```
Comment (kolyshkin) [inline tests/integration/env.bats:81]: This is not entirely correct. `test -n` checks if the string is empty, it's `test -v` that checks if the variable is set.

And, of course, empty value does not make a variable unset, and it shouldn't.

File: libcontainer/env.go (hunk 1 above)
Comment (rata) [reply libcontainer/env.go]: You are right, it does matter the duplication in the env. I realized later when [this docker test[[link] was failing. But forgot to update here, sorry!

If you add to the env: `"HOME=", HOME=foo"` that check in the docker test fails. I've also tested that `exec()` doesn't de-dup the env and then just running `env` shows all the vars. So I added a test for that too. I think there might be other improvements I can do to the tests, let me see.

However, I don't think we want the behavior you show in the diff. That is not how runc was behaving in the past (see runc 1.2, for example). If you give that env to runc 1.2, it will set the home reading `/etc/`. I've just tested with this bats:

```
@test "kir example" {
        update_config ' .process.env += ["HOME=/doo", "HOME="]'
        update_config ' .process.args = ["env"]'

        runc run test_busybox
        [ "$status" -eq 0 ]
}
```
You can run it with `--show-output-of-passing-tests`. You will see runc sets HOME to /root and only one HOME var is in the env with runc 1.2.

So, I think we do need to not add duplicates to the env, but we don't need to mark home as unseen and explore other settings of home. If the last HOME setting is unset, then we set home only once to what etc says.

I'll also add tests about no duplicates, now that I know it is important :)

[link]: https://github.com/moby/moby/blob/843e51459f14ebc964d349eba1013dc8a3e9d52e/integration-cli/docker_cli_run_test.go#L846-L847

File: tests/integration/env.bats (hunk 3 above)
Comment (rata) [reply tests/integration/env.bats:63]: Fair point. I'll update it, but now that I know the shell handles duplicates in weird ways, I'm not sure if I'll use `test` or not. I'll update soon :)

File: libcontainer/env.go (hunk 4 above)
Comment (rata) [reply libcontainer/env.go:63]: Fixed, thanks!

File: tests/integration/env.bats (hunk 5 above)
Comment (rata) [reply tests/integration/env.bats]: Oh, is it a built-in? I don't see `-v` in my test manpage.

But with the env duplication thing we discussed, I removed this test. I'm testing it's only set once and then that is set to the right value now (see "empty env var override is set only once" and "env var override")

File: libcontainer/env.go (hunk 1 above)
Comment (rata) [reply libcontainer/env.go]: Added: "env var HOME is set only once" and "env var override is set only once"  to check for duplicates.

The test "empty HOME env var is overridden with multiple overrides", that I think was there before (?), is checking basically that the home dir is not marked as unseen if it's set as empty.

This is the behavior that runc 1.2 also has (even though there is no "seen" there).


Let me know what you think

**Comment by rata** [comment]
@kolyshkin Fixed all the comments and added unit tests. PTAL. I left unresolved the only comment that I think the discussion is interesting and you might want to see in more detail.

File: tests/integration/env.bats (hunk 3 above)
Comment (rata) [reply tests/integration/env.bats]: I ended up changing it to use the env cmd, instead of test. See "env var with new-line is honored"

File: tests/integration/env.bats [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+	# There should be 4 lines
+	# NEW_LINE is a \n and when printed, it takes another line:
+	# 1. HOME=...
+	# 2. PATH=...
+	# 3. NEW_LINE_ENV=
+	# 4.
+	#
+	line_nr=$(echo "$output" | wc -l)
+	[[ $line_nr -eq 4 ]]
```
Comment (kolyshkin) [inline tests/integration/env.bats:93]: nit: can be simplified to
```bash
[ "$(wc -l <<<"$output")" -eq 4 ]
```

but it's not important.

File: tests/integration/env.bats [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+	# env will show if an env var is set multiple times.
+	update_config ' .process.args = ["env"]'
+	update_config ' .process.env = ["ONE=two", "ONE=", "PATH=/usr/bin:/bin"]'
+
+	runc run test_busybox
+	[ "$status" -eq 0 ]
+
+	# There should be 3 words/env-vars: ONE, PATH and HOME.
+	var_count=$(echo "${output}" | wc -w)
+	[[ $var_count -eq 3 ]]
```
Comment (kolyshkin) [inline tests/integration/env.bats:66]: nit: can be simplified to
```
[ "$(wc -w <<<"$output")" -eq 3 ]
```
but it's not important

File: tests/integration/env.bats [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+	# env will show if an env var is set multiple times.
+	update_config ' .process.args = ["env"]'
+	update_config ' .process.env = ["HOME=", "PATH=/usr/bin:/bin"]'
+
+	runc run test_busybox
+	[ "$status" -eq 0 ]
+
+	# There should be 2 words/env-vars: HOME and PATH.
+	var_count=$(echo "${output}" | wc -w)
+	[[ $var_count -eq 2 ]]
```
Comment (kolyshkin) [inline tests/integration/env.bats:53]: nit: can be simplified to
```
[ "$(wc -w <<<"$output")" -eq 2 ]
```
but it's not important

File: tests/integration/env.bats [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+	update_config ' .process.args += ["-c", "echo $HOME"]'
+
+	runc run test_busybox
+	[ "$status" -eq 0 ]
+	[[ "${lines[0]}" == '/override' ]]
+}
+
+@test "empty HOME env var is overridden" {
+	update_config ' .process.env += ["HOME="]'
+	update_config ' .process.args = ["env"]'
```
Comment (kolyshkin) [inline tests/integration/env.bats:27]: nit: maybe use `echo $HOME` here like in other adjacent tests.

**Review by kolyshkin — APPROVED** [review]
LGTM except for a few nits (feel free to ignore). Thanks for testing runc 1.2, too, this is important.

It would be nice to add failed Docker tests links (those from this PR description) to the second commit message and/or to bats file comment.

Thanks!

**Review by kolyshkin — COMMENTED** [review]
Almost forgot: please fix `prepareEnv` docstring accordingly, i.e. something like:

```diff
-//   - adds HOME to returned environment, if not found in the list.
+//   - adds HOME to returned environment, if not found in the list,
+//     or the value is empty.
```

File: libcontainer/env.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/env.go]: Yes, makes sense.

Even if it doesn't make sense, this is what runc used to do, and we have to keep this behavior to keep maintaining backward compatibility. I mean, we can break backward compatibility if there's a compelling reason to do so, but I don't see why we should do it here.

So, I agree, it should work the same way as runc 1.2 (or runc 1.0, etc.)

Comment (rata) [reply libcontainer/env.go]: Thanks!

File: tests/integration/env.bats (hunk 6 above)
Comment (rata) [reply tests/integration/env.bats:93]: Fixed, thanks!

File: tests/integration/env.bats (hunk 7 above)
Comment (rata) [reply tests/integration/env.bats:66]: Fixed, thanks!

File: tests/integration/env.bats (hunk 8 above)
Comment (rata) [reply tests/integration/env.bats:53]: Fixed, thanks!

**Comment by rata** [comment]
@kolyshkin thanks, fixed!

**Comment by kolyshkin** [comment]
Still LGTM, thanks!

**Comment by rata** [comment]
@AkihiroSuda @lifubang PTAL

File: tests/integration/env.bats [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	[ "$status" -eq 0 ]
+
+	# There should be 4 lines
+	# NEW_LINE is a \n and when printed, it takes another line:
+	# 1. HOME=...
+	# 2. PATH=...
+	# 3. NEW_LINE_ENV=
+	# 4.
+	#
+	[ "$(wc -l <<<"$output")" -eq 4 ]
```
Comment (AkihiroSuda) [inline tests/integration/env.bats:96]: It might be more robust to check the actual string, not just length

**Review by AkihiroSuda — APPROVED** [review]

**Comment by kolyshkin** [comment]
1.3 backport done (included into 1.3.0-rc.2): https://github.com/opencontainers/runc/pull/4711

Fixing the labels accordingly.
